### ISSUE TITLE
Running the code

### ISSUE THREAD

**Issue by darsh10** [Shivanshu-Gupta/Pytorch-POS-Tagger#1]
Hi,

The postagger.py file is missing. What is the format of the data, train_data.tsv ?

**Comment by Shivanshu-Gupta** [comment]
Ohh it should have been main.py instead of postagger.py. I have corrected the README.md and added steps for generating the tsvs. It should work now. Also, although most arguments are taken from command-line, some paths may have been hardcoded in the code itself, so be on a lookout!

**Comment by darsh10** [comment]
Thanks! The code is running now. However, the performance on the validation set is 0.

**Comment by Shivanshu-Gupta** [comment]
You mean validation accuracy is being reported as 0? What exactly did you try doing? If possible please share the output as well.

**Comment by cswangjiawei** [comment]
In the 'train()' function in 'train.py', there is 'loss = running_loss / example_count', and the loss function is 'criterion = nn.NLLLoss()', 'size_average' defaults to True. If 'loss = running_loss / example_count', should the loss function be 'criterion = nn.NLLLoss(size_average=False)'?

**Comment by Shivanshu-Gupta** [comment]
You are right. The average loss is being computed incorrectly. That said, the model is still being trained correctly, as it's only being computed to track the training progress via logging.

**Fix**: One way to fix it would be to set size_average=False as you suggested. But that would change the loss function and affect training. The right way would be to change `running_loss += loss.data[0]` to `running_loss += loss.data[0] * length * (rng[1] - rng[0])`. Here `length * (rng[1] - rng[0])` gives you the #words tagged in a mini-batch. In the end, by dividing by `example_count` which is the total #words tagged in the epoch, you will get the correct average loss.

P.S.: Check out [Hierarchical Attention Network's train script](https://github.com/Shivanshu-Gupta/hierarchical-attention-network/blob/master/train.py) for a more modular implementation of Tracker that helps prevent such mistakes! ;)

**Comment by cswangjiawei** [comment]
In "text_vocab_size = len (text.vocab.stoi) + 1", "text.vocab"  includes all of the words and ' <unk> ',' <pad> ', Why' len (text.vocab.stoi) 'Add 1 more?

**Comment by cswangjiawei** [comment]
In "text_vocab_size = len (text.vocab.stoi) + 1", "text.vocab" includes all of the words and "<unk>" ,"<pad>" , Why' len (text.vocab.stoi) 'Add 1 more?

**Comment by cswangjiawei** [comment]
In "text_vocab_size = len (text.vocab.stoi) + 1", "text.vocab" includes all of the words and 'unk ','pad ', Why' len (text.vocab.stoi) 'Add 1 more?

**Comment by Shivanshu-Gupta** [comment]
It was required for the specific version of torchtext that I was using at that time. In that version, the 'unk' token was not added as a part of `build_vocab()` call. It got added later when a token wasn't found in the vocabulary. This I think changed in the later versions. Just check what vocabulary `build_vocab()` is generating for you and accordingly set the values of `text_vocab_size` and `tag_vocab_size`. Also, remember you don't want to count 'unk' or 'pad' in `tag_vocab_size`. Let me know if you have any issues.
